tamr_client/datasets/dataframe.py

```diff
@@ -0,0 +1,36 @@
+"""
+Convenient functionality for interacting with pandas DataFrames.
+"""
+
+import json
+
+import pandas as pd
+
+import tamr_client as tc
+from tamr_client.types import JsonDict
+
+
+def upsert(
+    session: tc.Session, dataset: tc.Dataset, df: pd.DataFrame, *, primary_key_name: str
+) -> JsonDict:
+    """Upserts a record for each row of `df` with attributes for each column in `df`.
+    Args:
+        dataset: Dataset to receive record updates
+        df: The DataFrame containing records to be upserted
+        primary_key_name: The primary key of the dataset.  Must be a column of `df`.
+
+    Returns:
+            JSON response body from the server
+
+    Raises:
+        requests.HTTPError: If an HTTP error is encountered
+        PrimaryKeyNotFound: If `primary_key_name` is not a column in `df`.
+    """
+    if primary_key_name not in df.columns:
+        raise tc.PrimaryKeyNotFound(
+            f"Primary key: {primary_key_name} is not in DataFrame column names: {df.columns}"
+        )
+    records = json.loads(df.to_json(orient="records"))
```

Comment:
this will re-materialize the entire dataframe as a single string, which will definitely be a memory issue, and will almost certainly also be a performance issue

Reply:
Consider using [`iterrrows`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iterrows.html) instead of `to_json` to avoid memory issues

Reply:
I think we can address that in the function by stringifying and uploading in batches (1000? rows at a time).

A related feature that we've previously discussed is handling the large DataFrame problem with an `upsert_from_csv` function that would cover the issue of not even being able to load everything into a single DataFrame for very large sets of data.

Reply:
There are two issues with uploading large datasets:

1. Loading the entire dataset into memory can cause OOM errors.
2. After the client is done sending the data but is still waiting for the response from the server, the server may take too long processing the data. If too much time elapses without data being sent over the connection, the connection will be severed before the server has a chance to send its response.

(1) should be easy to avoid by doing streaming via `iterrows`
(2) might be resolved via chunking, but we should first implement a solution for (1) and then reproduce issue (2) before trying to solve (2)

Reply:
My main motivation for using `.to_json()` is the handling of changing `nan` to `null` now that we aren't using `simplejson`.

I could see incorporating this in the following way:
```python
records = map(lambda x: x[1].to_json(), df.iterrows())
```

I'm interested if there are issues (functionally it seems okay to me, but codestyle seems iffy).

Reply:
Only nit is that for consistency, we should use a generator expression rather than `map`:

```python
records = (x[1].to_json() for x in df.iterrows())
```

Reply:
One issue I've found with this is that each record becomes a JSON-string instead of a dict.  I can handle this by instead using:
```python
records = (json.loads(x[1].to_json()) for x in df.iterrows())
```
which is particularly ugly.
To be clear, what we gain from the above as opposed to 
```python
records = (x[1].to_dict() for x in df.iterrows())
```
is the handling of `np.nan`, without using `simplejson` or writing our own json encoder.

Reply:
One option is that you could separate the steps to make a generator pipeline (for readability):

```python
# serialize records via to_json to handle `np.nan` values
serialized_records = (x[1].to_json() for x in df.iterrows())
records = (json.loads(x) for x in serialized_records)
```